apps/sync/js/sync-engine/syncengine.js

```diff
@@ -0,0 +1,274 @@
+      }
+      // See https://docs.services.mozilla.com/storage/apis-1.5.html
+      return window.btoa(binStr).replace('+', '-').replace('/', '_');
+    },
+    validate: function(id) {
+      // FxSync id's should be 12 ASCII characters, representing 9 bytes of data
+      // in modified Base64 for URL variants exist, where the '+' and '/'
+      // characters of standard Base64 are respectively replaced by '-' and '_'
+      // See https://docs.services.mozilla.com/storage/apis-1.5.html
+      return /^[A-Za-z0-9-_]{12}$/.test(id);
+    }
+  });
+
+  var ControlCollectionIdSchema = Kinto.createIdSchema({
+    constructor: function(collectionName, keyName) {
+      this.collectionName = collectionName;
+      this.keyName = keyName;
+    },
+    generate: function() {
+      return this.keyName;
+    },
+    validate: function(id) {
+      return (id === this.keyName);
+    }
+  });
+
+  var WebCryptoTransformer = Kinto.createRemoteTransformer({
+    constructor: function(collectionName, fswc) {
+      if (!fswc.bulkKeyBundle) {
+        throw new Error(`Attempt to register Transformer with no bulk key bundl\
+e!`);
+      }
+      this.collectionName = collectionName;
+      this.fswc = fswc;
+    },
+    encode: function(record) {
+      return this.fswc.encrypt(record.payload, this.collectionName)
+      .then(payloadEnc => {
+        record.payload = JSON.stringify(payloadEnc);
+        return record;
+      });
+    },
+    decode: function(record) {
+      // Allowing JSON.parse errors to bubble up to the errors list in the
+      // syncResults:
+      return this.fswc.decrypt(JSON.parse(record.payload), this.collectionName)
+      .then(payloadDec => {
+        record.payload = payloadDec;
+        return record;
+      });
+    }
+  });
+
+  var SyncEngine = function(options) {
+    if (typeof options !== 'object') {
+      throw new Error('options should be an Object');
+    }
+    if (typeof options.URL !== 'string') {
+      throw new Error('options.URL should be a String');
+    }
+    if (typeof options.assertion !== 'string') {
+      throw new Error('options.assertion should be a String');
+    }
+    if (typeof options.xClientState !== 'string') {
+      throw new Error('options.xClientState should be a String');
+    }
+    if (typeof options.kB !== 'string') {
+      throw new Error('options.kB should be a String');
+    }
+
+    this._kB = options.kB;
+    this._collections = {};
+    this._controlCollections = {};
+    this._adapters = {};
+    this._fswc = new FxSyncWebCrypto();
+    this._initKinto({
+       URL: options.URL,
+       assertion: options.assertion,
+       xClientState: options.xClientState
+    });
+  };
+
+  SyncEngine.prototype = {
+    _initKinto: function(kintoCredentials) {
+      this._kinto = new Kinto({
+        bucket: 'syncto',
+        remote: kintoCredentials.URL,
+        headers: {
+          'Authorization': 'BrowserID ' + kintoCredentials.assertion,
```

Comment:
If the assertion has expired, a 401 http response will come from the Syncto server, which will result in `SyncEngine#syncNow` rejecting its promise with a `SyncEngine.AuthError`. So then the sync app knows it needs to get a new assertion from SyncCredentials (not part of SyncEngine and not implemented yet).